src/streamer.rs

```diff
@@ -257,51 +53,115 @@ pub fn receiver(
     }))
 }
 
-fn recv_send(sock: &UdpSocket, recycler: &ResponseRecycler, r: &ResponseReceiver) -> Result<()> {
+fn recv_send(sock: &UdpSocket, recycler: &BlobRecycler, r: &BlobReceiver) -> Result<()> {
     let timer = Duration::new(1, 0);
-    let msgs = r.recv_timeout(timer)?;
-    let msgs_ = msgs.clone();
-    let mut num = 0;
-    msgs.read().unwrap().send_to(sock, &mut num)?;
-    recycle(recycler, msgs_);
+    let mut msgs = r.recv_timeout(timer)?;
+    Blob::send_to(recycler, sock, &mut msgs)?;
     Ok(())
 }
 
 pub fn responder(
     sock: UdpSocket,
     exit: Arc<AtomicBool>,
-    recycler: ResponseRecycler,
-    r: ResponseReceiver,
+    recycler: BlobRecycler,
+    r: BlobReceiver,
 ) -> JoinHandle<()> {
     spawn(move || loop {
-        if recv_send(&sock, &recycler, &r).is_err() && exit.load(Ordering::Relaxed) {
+        if recv_send(&sock, &recycler, &r).is_err() || exit.load(Ordering::Relaxed) {
             break;
         }
     })
 }
 
+//TODO, we would need to stick block authentication before we create the
+//window.
+fn recv_window(
+    window: &mut Vec<Option<SharedBlob>>,
+    recycler: &BlobRecycler,
+    consumed: &mut usize,
+    socket: &UdpSocket,
+    s: &BlobSender,
+) -> Result<()> {
+    let mut dq = Blob::recv_from(recycler, socket)?;
+    while let Some(b) = dq.pop_front() {
+        let b_ = b.clone();
+        let mut p = b.write().unwrap();
+        let pix = p.get_index()? as usize;
+        let w = pix % NUM_BLOBS;
+        //TODO, after the block are authenticated
+        //if we get different blocks at the same index
+        //that is a network failure/attack
+        {
+            if window[w].is_none() {
+                window[w] = Some(b_);
+            } else {
+                debug!("duplicate blob at index {:}", w);
+            }
+            //send a contiguous set of blocks
+            let mut dq = VecDeque::new();
+            loop {
```

Comment:
Looking like high cyclomatic complexity. Did this get by clippy?

Reply:
cippy doesn't run

Reply:
```
cargo +nightly clippy --features="unstable
lib: solana                                                                    
                                                                               
error: failed to run `rustc` to learn about target-specific information        
                                                                               
Caused by:                                                                
  process didn't exit successfully: `/Users/anatolyy/.cargo/bin/clippy-driver rustc - --crate-name ___ --print=file-names --target x86_64-apple-darwin --crate
-type bin --crate-type proc-macro --crate-type rlib` (signal: 6, SIGABRT: process abort signal)                                                               
--- stderr                                                                                                                                                    
dyld: Library not loaded: @rpath/librustc_driver-0d1bd24d5bdd5f98.dylib                                                                                       
  Referenced from: /Users/anatolyy/.cargo/bin/clippy-driver                                                                                                   
  Reason: image not found 
```

Reply:
You need to reinstall clippy any time you reinstall Rust nightly.